datafusion/common/src/dfschema.rs

```diff
@@ -420,6 +420,11 @@ impl DFSchema {
                 Self::datatype_is_semantically_equal(k1.as_ref(), k2.as_ref())
                     && Self::datatype_is_semantically_equal(v1.as_ref(), v2.as_ref())
             }
+            // The next two cases allow for the possibility that one schema has a dictionary encoded array
```

Comment:
I am somewhat worried about this change (as it effectively changes the meaning of the function to also include logical types).

I think we should either  update the comments to reflect this change, or (my preference) make a new function that is explicit. Perhaps something like :

```rust
   ///  Returns true of two [`DataType`]s have the same logical data type:
   /// Either the same name and type, or if one is dictionary encoded, then 
   /// the same name and type of the values.
   /// E.g. Dictionary(_, Utf8) is semantically equivalent to Utf8 since both represent an array of strings
  fn datatype_has_same_logical_type(
        match (dt1, dt2) {
            (DataType::Dictionary(_, v1), othertype) => v1.as_ref() == othertype,
            (othertype, DataType::Dictionary(_, v1)) => v1.as_ref() == othertype,
             _ => Self::datatype_is_semantically_equal(dt1, dt2)
        }
  }
```

Reply:
I think your concern is justified as the optimizer also relies on this function and might have stricter equivalence requirements. I created a separate method for logical equivalence which allows for different dictionary encodings as long as values can ultimately be resolved to the same type.

The optimizer continues to use the original semantic equivalence method, and I've updated the insert_into methods to use the softer logical equivalence check.